pkg/controller/replica.go

```diff
@@ -0,0 +1,338 @@
+	"sigs.k8s.io/azuredisk-csi-driver/pkg/azureutils"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	deletionPollingInterval = time.Duration(10) * time.Second
+)
+
+type ReconcileReplica struct {
+	client                client.Client
+	azVolumeClient        azVolumeClientSet.Interface
+	kubeClient            kubeClientSet.Interface
+	namespace             string
+	controllerSharedState *SharedState
+	deletionMap           sync.Map
+	cleanUpMap            sync.Map
+	mutexLocks            sync.Map
+}
+
+var _ reconcile.Reconciler = &ReconcileReplica{}
+
+func (r *ReconcileReplica) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	azVolumeAttachment, err := azureutils.GetAzVolumeAttachment(ctx, r.client, r.azVolumeClient, request.Name, request.Namespace, true)
+	if errors.IsNotFound(err) {
+		klog.Infof("AzVolumeAttachment (%s) has been successfully deleted.", request.Name)
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		klog.Errorf("failed to fetch AzVolumeAttachment (%s): %v", request.Name, err)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	if azVolumeAttachment.Spec.RequestedRole == v1alpha1.PrimaryRole {
+		// Deletion Event
+		if deletionRequested(&azVolumeAttachment.ObjectMeta) {
+			// If primary attachment is marked for deletion, queue garbage collection for replica attachments
+			r.triggerGarbageCollection(azVolumeAttachment.Spec.UnderlyingVolume)
+		} else {
+			// If not, cancel scheduled garbage collection if there is one enqueued
+			r.cancelGarbageCollection(azVolumeAttachment.Spec.UnderlyingVolume)
+
+			// If promotion event, create a replacement replica
+			if isAttached(azVolumeAttachment) && azVolumeAttachment.Status.Detail.Role != azVolumeAttachment.Spec.RequestedRole {
+				if err := r.manageReplicas(ctx, azVolumeAttachment.Spec.UnderlyingVolume); err != nil {
+					return reconcile.Result{Requeue: true}, err
+				}
+			}
+		}
+	} else {
+		// create a replacement replica if replica attachment failed or promoted
+		if deletionRequested(&azVolumeAttachment.ObjectMeta) {
+			if azVolumeAttachment.Annotations == nil || !metav1.HasAnnotation(azVolumeAttachment.ObjectMeta, azureutils.CleanUpAnnotation) {
+				go func() {
+					conditionFunc := func() (bool, error) {
+						var tmp v1alpha1.AzVolumeAttachment
+						err := r.client.Get(ctx, request.NamespacedName, &tmp)
+						if errors.IsNotFound(err) {
+							return true, nil
+						}
+
+						return false, err
+					}
+					_ = wait.PollImmediateInfinite(deletionPollingInterval, conditionFunc)
```

Comment:
Should we also set a timeout here instead of waiting infinitely?

Reply:
In this case, we are using a separate goroutine to handle deletion so I think `PollImmediateInfinite` makes sense here. Also, if we requeue after a timeout, there is a chance that the reconciler does not create a replacement replica attachment (if CRI's finalizer is removed in between poll timeout and next reconciliation).